language/move-vm/runtime/src/data_cache.rs

```diff
@@ -89,54 +82,57 @@ impl<'r, 'l, R: RemoteCache> TransactionDataCache<'r, 'l, R> {
+            let mut modules = BTreeMap::new();
+            for (module_name, module_blob) in account_data_cache.module_map {
+                modules.insert(module_name, Some(module_blob));
+            }
+
+            let mut resources = BTreeMap::new();
+            for (ty, (layout, gv)) in account_data_cache.data_map {
                 match gv.into_effect()? {
                     GlobalValueEffect::None => (),
                     GlobalValueEffect::Deleted => {
-                        if let TypeTag::Struct(s_tag) = self.loader.type_to_type_tag(&ty)? {
-                            vals.push((s_tag, None))
-                        } else {
-                            // non-struct top-level value; can't happen
-                            return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR));
-                        }
+                        let struct_tag = match self.loader.type_to_type_tag(&ty)? {
+                            TypeTag::Struct(struct_tag) => struct_tag,
+                            _ => return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)),
+                        };
+                        resources.insert(struct_tag, None);
                     }
                     GlobalValueEffect::Changed(val) => {
-                        if let TypeTag::Struct(s_tag) = self.loader.type_to_type_tag(&ty)? {
-                            vals.push((s_tag, Some((ty_layout, val))))
-                        } else {
-                            // non-struct top-level value; can't happen
-                            return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR));
-                        }
+                        let struct_tag = match self.loader.type_to_type_tag(&ty)? {
+                            TypeTag::Struct(struct_tag) => struct_tag,
+                            _ => return Err(PartialVMError::new(StatusCode::INTERNAL_TYPE_ERROR)),
+                        };
+                        let resource_blob = val.simple_serialize(&layout).ok_or_else(|| {
+                            PartialVMError::new(StatusCode::VALUE_SERIALIZATION_ERROR)
+                        })?;
+                        resources.insert(struct_tag, Some(resource_blob));
                     }
                 }
             }
-            if !vals.is_empty() {
-                resources.push((addr, vals));
-            }
-            modules.extend(
-                account_cache
-                    .module_map
-                    .into_iter()
-                    .map(|(module_id, blob)| (module_id, blob)),
-            );
+
+            account_changesets.insert(addr, AccountChangeSet { modules, resources });
         }
 
         let mut events = vec![];
         for (guid, seq_num, ty, ty_layout, val) in self.event_data {
             let ty_tag = self.loader.type_to_type_tag(&ty)?;
-            events.push((guid, seq_num, ty_tag, ty_layout, val))
+            // REVIEW: This error doesn't look right.
```

Comment:
Agreed; this should be something that results in a discard. Perhaps `INTERNAL_TYPE_ERROR`? I think that's the only reason this can fail--`ty_layout` doesn't match `val`.

Reply:
Should this be classified as a breaking change? I mean we're switching to a different bucket here, but this should be dead code if the VM isn't bugged...

Reply:
I wouldn't worry about this case being a breaking change or not. But should it be a VALUE_SERIALIZATION_ERROR? that matches whats done for the rest of the changeset

Reply:
A bigger question is should these two serialization errors be user errors or invariant violations? Maybe FAILED_TO_SERIALIZE_XXX or something?

Reply:
I would vote for invariant violation -- both the value and the layout are generated by the Move VM itself and there's no user action that could cause this to fail.

Reply:
I definitely think this should be an invariant violation, but don't have a strong opinion about the specific error code.